seed-job/index_reader.py

```diff
@@ -0,0 +1,369 @@
+            with open(filepath) as fin:
+                data = yaml.load(fin, Loader=yaml.BaseLoader)
+        except yaml.YAMLError as exc:
+            raise(exc)
+        else:
+            return data
+
+    def read_projects(self):
+        """
+        Reads the projects from given container index
+        """
+        projects = []
+
+        for yamlfile in glob(self.index + "/*.y*ml"):
+            # skip index_template
+            if "index_template" in yamlfile:
+                continue
+
+            app = self.read_yaml(yamlfile)
+            for entry in app['Projects']:
+                # create a project object here with all properties
+                project = Project(entry, self.namespace)
+                # append to the list of projects
+                projects.append(project)
+
+        return projects
+
+
+class DeploymentConfigManager(object):
+    """
+    This class represents utilities to manage
+    deployment configs on openshift as used
+    by pipeline service
+    """
+
+    def __init__(self, registry_url, namespace, from_address, smtp_server):
+        self.registry_url = registry_url
+        self.namespace = namespace
+        self.from_address = from_address
+        self.smtp_server = smtp_server
+        self.seed_template_params = """\
+-p GIT_URL={git_url} \
+-p GIT_PATH={git_path} \
+-p GIT_BRANCH={git_branch} \
+-p TARGET_FILE={target_file} \
+-p BUILD_CONTEXT={build_context} \
+-p DESIRED_TAG={desired_tag} \
+-p DEPENDS_ON={depends_on} \
+-p NOTIFY_EMAIL={notify_email} \
+-p PIPELINE_NAME={pipeline_name} \
+-p APP_ID={app_id} \
+-p JOB_ID={job_id} \
+-p PRE_BUILD_CONTEXT={pre_build_context} \
+-p PRE_BUILD_SCRIPT={pre_build_script} \
+-p REGISTRY_URL={registry_url} \
+-p FROM_ADDRESS={from_address} \
+-p SMTP_SERVER={smtp_server}"""
+
+    def list_all_buildConfigs(self):
+        """
+        List all available buildConfigs
+        returns list of buildConfigs available
+        """
+        command = "oc get bc -o name -n {}".format(self.namespace)
+        bcs = run_cmd(command)
+        if not bcs.strip():
+            return []
+        else:
+            return bcs.strip().split("\n")
+
+    def oc_apply_seedjob_template_command(
+            self, template_location="seed-job/template.yaml"):
+        """
+        Returns the oc process and oc apply commands with parameters
+        to process seed-job/template.yaml
+        """
+        oc_apply = "oc apply -n {} -f -".format(self.namespace)
+        oc_process = "oc process -f {0} {1}".format(
+            template_location,
+            self.seed_template_params
+        )
+        return oc_process + "|" + oc_apply
+
+    def apply_buildconfigs(self, project):
+        """
+        Given a project object representing a project in container index,
+        process the seed-job template for same and oc apply changes
+        """
+        # get the oc process and oc apply command
+        command = self.oc_apply_seedjob_template_command()
```

Comment:
Keeping a separate function for this seems a bit of overkill. We do the processing right here. Or maybe have the `command` set in `__init__` function. A separate function call for every project that we want to create buildconfig for just to create the command that needs to be executed doesn't seem reasonable.

Reply:
I don't think its an overkill TBH. They are methods of the same class and instead of having long methods with multiple operations, I separated the methods, the possible contant values are already set in `__init__`, for eg: `self.seed_template_params`. It's just fine too, having the command in same method, but I think the method will have a bunch of processing.
And say, if we need to update the index reader or some commands in it, we have one method to look into.

I am open to change it, its just that, how we want our methods to look like, with small separate utilities or with one method with a bunch of operations.

Also, I wanted to understand, every entry in the index would need processing via multiple methods, why calling an additional method would be unreasonable?

Reply:
> instead of having long methods with multiple operations, I separated the methods

I think long methods should be OK as long as the steps in it are relevant to what the method is trying to do. It also helps the future contributors to read through it one go instead of shuttling between functions. 

If the command is going to stay the same, we can have a `self.command` set in `__init__()`, no?

> And say, if we need to update the index reader or some commands in it, we have one method to look into.

We still have to update at one place only if we move this into the same function or `__init__`, no?

> small separate utilities or with one method with a bunch of operations

Bunch of operations sound OK as long as they are relevant to the overall goal of the method.

> why calling an additional method would be unreasonable

I'm looking at this only from number of function calls point of view. Also passing data between functions while it can be processed in the same.

Reply:
I mean, its fine keeping the operation in same method, its just that, I thought separating it out in separate methods would be more readable. As long as, team finds it comfortable, I am fine. Merging the two methods into one.

Reply:
updated the PR with discussed changes